src/jrd/BlobUtil.cpp

```diff
@@ -0,0 +1,281 @@
+/*
+ *  The contents of this file are subject to the Initial
+ *  Developer's Public License Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the
+ *  License. You may obtain a copy of the License at
+ *  http://www.ibphoenix.com/main.nfs?a=ibphoenix&page=ibp_idpl.
+ *
+ *  Software distributed under the License is distributed AS IS,
+ *  WITHOUT WARRANTY OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing rights
+ *  and limitations under the License.
+ *
+ *  The Original Code was created by Adriano dos Santos Fernandes
+ *  for the Firebird Open Source RDBMS project.
+ *
+ *  Copyright (c) 2020 Adriano dos Santos Fernandes <adrianosf@gmail.com>
+ *  and all contributors signed below.
+ *
+ *  All Rights Reserved.
+ *  Contributor(s): ______________________________________.
+ */
+
+#include "firebird.h"
+#include "../jrd/BlobUtil.h"
+#include "../jrd/blb.h"
+#include "../jrd/tra.h"
+
+using namespace Jrd;
+using namespace Firebird;
+
+
+namespace
+{
+	blb* getBlobFromHandle(thread_db* tdbb, ISC_INT64 handle)
+	{
+		const auto transaction = tdbb->getTransaction();
+		blb* blob;
+
+		if (transaction->tra_blob_util_map.get(handle, blob))
+			return blob;
+		else
+		{
+			status_exception::raise(Arg::Gds(isc_invalid_blob_util_handle));
+			return nullptr;
+		}
+	}
+}
+
+namespace Jrd {
+
+//--------------------------------------
+
+IExternalResultSet* BlobUtilPackage::appendProcedure(ThrowStatusExceptionWrapper* status,
+	IExternalContext* context, const AppendInput::Type* in, void*)
+{
+	const auto tdbb = JRD_get_thread_data();
+	Attachment::SyncGuard guard(tdbb->getAttachment(), FB_FUNCTION);
```

Comment:
It could be avoided for simple cases. Do you mean something different?

Reply:
> It could be avoided for simple cases. Do you mean something different?

I was been talking about avoinding SyncGuard for simple cases.

But yes, we should better avoid EngineCheckout for system packages in general.